packages/generator-joplin/generators/app/templates/api/types.ts

```diff
@@ -533,6 +533,43 @@ export interface MarkdownItContentScriptModule extends Omit<ContentScriptModule,
 	plugin: (markdownIt: any, options: any)=> any;
 }
 
+type EditorCommandCallback = (...args: any[])=> any;
+
+export interface CodeMirrorControl {
+	/** Points to a CodeMirror 6 EditorView instance. */
+	editor: any;
+	cm6: any;
+
+	/** `extension` should be a [CodeMirror 6 extension](https://codemirror.net/docs/ref/#state.Extension). */
+	addExtension(extension: any|any[]): void;
+
+	supportsCommand(name: string): boolean;
+	execCommand(name: string, ...args: any[]): any;
+	registerCommand(name: string, callback: EditorCommandCallback): void;
+
+	joplinExtensions: {
+		/**
+		 * Returns a [CodeMirror 6 extension](https://codemirror.net/docs/ref/#state.Extension) that
+		 * registers the given [CompletionSource](https://codemirror.net/docs/ref/#autocomplete.CompletionSource).
+		 *
+		 * Use this extension rather than the built-in CodeMirror [`autocompletion`](https://codemirror.net/docs/ref/#autocomplete.autocompletion)
+		 * if you don't want to use [languageData-based autocompletion](https://codemirror.net/docs/ref/#autocomplete.autocompletion^config.override).
+		 *
+		 * Using `autocompletion({ override: [ ... ]})` causes errors when done by multiple plugins.
+		 */
+		completionSource(completionSource: any): any;
+
+		/**
+		 * Creates an extension that enables or disables [`languageData`-based autocompletion](https://codemirror.net/docs/ref/#autocomplete.autocompletion^config.override).
+		 */
+		enableLanguageDataAutocomplete: { of: (enabled: boolean)=> any };
+	};
+}
+
+export interface CodeMirrorContentScriptModule extends Omit<ContentScriptModule, 'plugin'> {
```

Comment:
```suggestion
export interface MarkdownEditorContentScriptModule extends Omit<ContentScriptModule, 'plugin'> {
```

(**Optional**): It may make sense to rename `CodeMirrorContentScriptModule` to `MarkdownEditorContentScriptModule`. This would be more consistent with naming in the documentation and changelog.

If this is done, it might also make sense to rename `CodeMirrorControl` to `MarkdownEditorControl`.

Reply:
I agree, but won't that cause a regression for those who are using the existing name?

Reply:
`CodeMirrorContentScriptModule` hasn't been released as a part of `generator-joplin` yet — users would need to have copied either the `codemirror6/` example plugin or directly from `lib/`.

While this is possible, I find it unlikely — the [plugin tutorial](https://joplinapp.org/help/api/tutorials/cm6_plugin) does not use or suggest copying this type and the CodeMirror 6 plugin was updated only recently to use this type.

Reply:
Ok that sounds good then, let's use `MarkdownEditorContentScriptModule`

Reply:
See https://github.com/laurent22/joplin/pull/10015